### ISSUE TITLE
Allow default value of basic type in generic function without subtype constraint

### ISSUE THREAD

**Issue by SnowingFox** [microsoft/TypeScript#54597]
# Suggestion

## 🔍 Search Terms

```ts
const foo = <T extends number> (num: T = 1) => {
    return num
}
```

it will throw an error
```bash
Type 'number' is not assignable to type 'T'.
  'number' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'number'
```

we should fix this problem like this
```ts
num: T = 1 as T
```

but number is an basic type, it doesn't has subtype, so this type of error reporting is very confusing and carries a mental burden

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

i think compiler shouldn't throw an error in this case
## 💻 Use Cases

Reduce mental burden and make the world a better place

```[tasklist]
### Tasks
```

**Comment by fatcerberus** [comment]
> number is an basic type, it doesn't has subtype

Not true.  ‘2 | 3 | 4‘ is a subtype of ‘number‘ and ‘1‘ isn’t assignable to it.

**Comment by CwRv07** [comment]
I also found this problem in my case as follows

[typescript playground](https://www.typescriptlang.org/play?ts=5.1.3#code/PQKgUABBCyCeEFEB2BzANgSwM4AsLYiQHsAXCFIogEwgEMkaBJCAW1vh1oDcBTV7AE48ADkKw8kJHjRI4+AMyJo0RAO4ZUgLH+wkCM1X0yJIhACu48hJ4CMAYyz4kEIgKrWIxiPNq2MmErRSHnIesMJ8trRotqZogRhETkTyEEIkpgJOXFGmPFgANBC69EwQBpIeJraJWCQCtBp0DI4YJBhR2HyyfG7esSQOigKWSNZ2WDpQAEKmZMSqepZk9BDWAi4AXISmLABG7pFIxGT7dFhYGCijMibdI2O2EAAqTaWj0pVmFrQQWKa7tji52cKSQO32w08GlqhnaQSeugAyiZmJxeHQIABHXK1BJIQrYJAAcjILHgvVo-Qg2TQuRBEAAjBi-gCgQ5kts9tYAPyQEDAHQkMJ8OAAJR46UyT2FAB4ngA+CAAXmeAG4dPJTEhbG1El5KHLVgAPKQMBxgrnDFUM+UACgtWxe1oAlFsxRKMkhpeE5YqAN66NKezlgAC+YGqSFqHjyJAADMr9URbc6IzUjLGmSrFMmAEypnRgUC6RhOO4SUrVWI0RJoeA+HAYHjosl0DAsCC7WB8gWa7W6py7WgCQ08E0V83g9zWu0O56Jpm0BxPV0wWDiyVe2UKiABqBBzIh0NAA)

**Comment by SnowingFox** [comment]
> > number is an basic type, it doesn't has subtype
> 
> Not true. `2 | 3 | 4` is a subtype of `number` and `1` isn’t assignable to it. And if there were truly no subtypes then you wouldn’t need to write a generic function in the first place.

```ts
<T extends null> (num: T = null)
```

**Comment by fatcerberus** [comment]
@CwRv07 The reason this isn't allowed without a cast is

```ts
const x = bar<2>()  // claims to return 2 at compile time
//    ^?
console.log(x)  // prints 1 at runtime
```

[Playground](https://www.typescriptlang.org/play?ts=5.1.3#code/PQKgUABBCyCeEFEB2BzANgSwM4AsLYiQHsAXCFIogEwgEMkaBJCAW1vh1oDcBTV7AE48ADkKw8kJHjRI4+AMyJo0RAO4ZUgLH+wkCM1X0yJIhACu48hJ4CMAYyz4kEIgKrWIxiPNq2MmErRSHnIesMJ8trRotqZogRhETkTyEEIkpgJOXFGmPFgANBC69EwQBpIeJraJWCQCtBp0DI4YJBhR2HyyfG7esSQOigKWSNZ2WDpQAEKmZMSqepZk9BDWAi4AXISmLABG7pFIxGT7dFhYGCijMibdI2O2EAAqTaWj0pVmFrQQWKa7tji52cKSQO32w08GlqhnaQSeugAyiZmJxeHQIABHXK1BJIQrYJAAcjILHgvVo-Qg2TQuRBEAAjBi-gCgQ5kts9tYAPyQEDAHQkMJ8OAAJR46UyT2FAB4ngA+CAAXmeAG4dPJTEhbG1EhBdrQBHLVgAPKQMBxgrnDFUM+UACitWxetrOzwAlFsxRKMkhpeE5YqAN66NK+zlgAC+OmqSFqEBNyv1hplACYHe6oMBgBBAQ0WA5PGHMhBU3QyNUWMI-F0MCweGBs1AoAA9XmxrBKHgAOhUKHtJszECbog0A0Z5dSWra9bAQA)

**Comment by MartinJohns** [comment]
You can type it `T | 1`.
